pypga/modules/migen/spimachine.py

```diff
@@ -1,250 +1,328 @@
+                # Shift next value into sdo on next system clock cycle
+                reg.shift.eq(1),
+                # Restart clock generator only, if there is another full cycle
+                NextValue(cg.start, self.counter > 1),
+                NextValue(self.counter, self.counter - 1),
+                NextState("1ST_HALFCYCLE")
+            ).Else(
+                # Continue current 2nd half cycle.
+                NextValue(cg.start, 0),
+            )
         )
-        fsm.act("POST",
-                # dummy half cycle before deasserting CS in CPHA=0
-                cg.count.eq(1),
-                self.cs_next.eq(1),
+        # Asserts CS for specified number of half cycles, while keeping clock and
+        # MOSI low.
+        fsm.act("TRAILING_HALFCYCLES",
+            self.cs.eq(1),
+            If(self.counter == 0,
+                # Proceed to idle half cycles.
+                # Start clock generator only, if there is at least one idle half cycle
+                NextValue(cg.start, idle_halfcycles > 0),
+                NextValue(self.counter, idle_halfcycles),
+                NextState("IDLE_HALFCYCLES")
+            ).Else(
                 If(cg.done,
-                    NextState("WAIT")
+                    # Restart clock generator only, if there is another trailing
+                    # half cycle.
+                    NextValue(cg.start, self.counter > 1),
+                    NextValue(self.counter, self.counter - 1),
+                    NextState("TRAILING_HALFCYCLES")
+                ).Else(
+                    # Continue current trailing half cycle.
+                    NextValue(cg.start, 0)
                 )
+            )
         )
-        fsm.act("WAIT",
-                # dummy half cycle to meet CS deassertion minimum timing
+        # Holds CS low for the specified number of half cycles.
+        fsm.act("IDLE_HALFCYCLES",
+            If(self.counter == 0,
+                # Proceed to idle state.
+                NextValue(cg.start, 0),
+                NextState("IDLE")
+            ).Else(
                 If(cg.done,
-                    NextState("IDLE")
+                    # Restart clock generator only, if there is another idle
+                    # half cycle.
+                    NextValue(cg.start, self.counter > 1),
+                    NextValue(self.counter, self.counter - 1),
+                    NextState("IDLE_HALFCYCLES")
                 ).Else(
-                    cg.count.eq(1),
-                    self.cs_next.eq(1)
+                    # Continue current idle half cycle.
+                    NextValue(cg.start, 0)
                 )
+            )
         )
-        # fsm.act("END",
-        #         cg.count.eq(1),
-        #         self.cs_next.eq(1),#1
-        #         If(cg.done,
-        #             NextState("IDLE")
-        #         )
-        # )
\ No newline at end of file
+
+
+def sim_spimachine(num_steps, query):
```

Comment:
This is a super convenient debugging tool to understand the temporal evolution of the different signals. I believe it should be quite doable to also set this up for `pypga` modules. I have some ideas regarding that and will hopefully come around to implementing this.

Reply:
Yes I agree. Please check the file `tests/unit/modules/migen/test_spi.py` - it does something similar. In general I suggest to keep tests in the `tests` folder, i.e. to move your test code starting from line 263 to a similar file (either also test_spi.py, or test_spi2.py etc). Then instead of running `python spimachine.py`, you'd rather run `pytest -s test_spi2.py`, and it would be possible to check assertions on every future commit, such that your code never breaks in the future.